src/bun.js/bindings/webcore/WebSocket.cpp

```diff
@@ -650,6 +650,211 @@ ExceptionOr<void> WebSocket::close(std::optional<unsigned short> optionalCode, c
     return {};
 }
 
+ExceptionOr<void> WebSocket::terminate()
+{
+    LOG(Network, "WebSocket %p terminate()", this);
+
+    if (m_state == CLOSING || m_state == CLOSED)
+        return {};
+    if (m_state == CONNECTING) {
+        m_state = CLOSING;
+        if (m_upgradeClient != nullptr) {
+            void* upgradeClient = m_upgradeClient;
+            m_upgradeClient = nullptr;
+            if (m_isSecure) {
+                Bun__WebSocketHTTPSClient__cancel(upgradeClient);
+            } else {
+                Bun__WebSocketHTTPClient__cancel(upgradeClient);
+            }
+        }
+        updateHasPendingActivity();
+        return {};
+    }
+    m_state = CLOSING;
+    switch (m_connectedWebSocketKind) {
+    case ConnectedWebSocketKind::Client: {
+        Bun__WebSocketClient__cancel(this->m_connectedWebSocket.client);
+        updateHasPendingActivity();
+        break;
+    }
+    case ConnectedWebSocketKind::ClientSSL: {
+        Bun__WebSocketClientTLS__cancel(this->m_connectedWebSocket.clientSSL);
+        updateHasPendingActivity();
+        break;
+    }
+    default: {
+        break;
+    }
+    }
+    this->m_connectedWebSocketKind = ConnectedWebSocketKind::None;
+    updateHasPendingActivity();
+    return {};
+}
+
+ExceptionOr<void> WebSocket::ping()
+{
+    auto message = WTF::String::number(WTF::jsCurrentTime());
+    LOG(Network, "WebSocket %p ping() Sending Timestamp '%s'", this, message.data());
+    if (m_state == CONNECTING)
+        return Exception { InvalidStateError };
+
+    // No exception is raised if the connection was once established but has subsequently been closed.
+    if (m_state == CLOSING || m_state == CLOSED) {
+        size_t payloadSize = message.length();
+        m_bufferedAmountAfterClose = saturateAdd(m_bufferedAmountAfterClose, payloadSize);
+        m_bufferedAmountAfterClose = saturateAdd(m_bufferedAmountAfterClose, getFramingOverhead(payloadSize));
+        return {};
+    }
+
+    // TODO: this->sendWebSocketString(message, Opcode::Ping);
```

Comment:
These are all stubbed out for now, mostly because there are framing bugs in `WebSocket` client that need to be resolved.